x/authz specs (#8499)
cosmos/cosmos-sdk: x/authz/spec/01_concepts.md

<!--
order: 1
-->

# Concepts

## Authorization

> Any concrete type of authorization defined in the `x/authz` module must fulfill `Authorization` interface outlined below. Authorizations determine exactly what privileges are granted. They are extensible and can be defined for any Msg service method even outside of the module where the Msg method is defined. Authorizations use the new `ServiceMsg` type from [ADR 031](docs/architecture/adr-31-msg-service.md).

Reviewer: ```suggestion
Any concrete type of authorization defined in the `x/authz` module must fulfill the `Authorization` interface outlined below. Authorizations determine exactly what privileges are granted. They are extensible and can be defined for any Msg service method even outside of the module where the Msg method is defined. Authorizations use the new `ServiceMsg` type from [ADR 031](../../../docs/architecture/adr-031-msg-service.md).
```

Author: I found relative path `../../../docs/architecture/adr-031-msg-service.md` showing 404 page on `https://docs.cosmos.network`.

Reviewer: ah, ok. How did you test that, btw?

Author: I ran this command `npm run serve` from  docs folder.  Same issue exists [here](https://docs.cosmos.network/v0.40/docs/architecture/adr-003-dynamic-capability-store.html).

> Any concrete type of authorization defined in the `x/authz` module must fulfill the `Authorization` interface outlined below. Authorizations determine exactly what privileges are granted. They are extensible and can be defined for any Msg service method even outside of the module where the Msg method is defined. Authorizations use the new `ServiceMsg` type from [ADR 031](../../../docs/architecture/adr-031-msg-service.md).

Reviewer 2: ```suggestion
Any concrete type of authorization defined in the `x/authz` module must fulfill the `Authorization` interface outlined below. Authorizations determine exactly what privileges are granted. They are extensible and can be defined for any Msg service method even outside of the module where the Msg method is defined. Authorizations use the new `ServiceMsg` type from [ADR 031](../../../architecture/adr-031-msg-service.md).
```
The ADR link was not workin when I rendered docs locally, this should fix it.

Reviewer 2: Unfortunately, we need to choose between having these docs files render proper links on **either** bowsing .md files on github OR when viewing on docs.cosmos.network. The two are incompatible when linking btw module & ADR docs as all the modules get copied into the `./docs` folder for building docs.cosmos.network. My preference is to have the links render correctly when viewed on docs.cosmos.network

Reviewer 2: The other option (if we don't want to do relative path), would be to have them link to absolute paths as rendered on docs.cosmos.network.

This could be done with `[ADR 031](/architecture/adr-031-msg-service.md)` which would always link to the default version of the docs (most recent released version), or we could hardcode a specific version and/or master like so:
`[ADR 031](/master/architecture/adr-031-msg-service.md)`.

My preference is for the relative path (as in my suggested change above).

> +++ https://github.com/cosmos/cosmos-sdk/blob/master/x/authz/types/authorizations.go#L15-L24

Reviewer: let's use a commit hash instead of master, as master's code can change (Same applies to all other code snippets)

## Built-in Authorizations

Cosmos-SDK `x/authz` module comes with following authorization types

### SendAuthorization

> `SendAuthorization` implements `Authorization` interface for the `MsgSend` ServiceMsg, that takes a `SpendLimit` and updates it down to zero.

Reviewer: ```suggestion
`SendAuthorization` implements `Authorization` interface for the `cosmos.bank.v1beta1.Msg/Send` ServiceMsg, that takes a `SpendLimit` and updates it down to zero.
```
